lsp-mode.el

```diff
@@ -7563,13 +7563,13 @@ When ARG is t the lsp mode will start new language server even if
 there is language server which can handle current language. When
 ARG is nil current file will be opened in multi folder language
 server if there is such. When `lsp' is called with prefix
-argument ask the user to select which language server to start. "
+argument ask the user to select which language server to start."
   (interactive "P")
 
   (when (and lsp-auto-configure (not lsp--client-packages-required))
     (seq-do (lambda (package)
               ;; loading client is slow and `lsp' can be called repeatedly
-              (unless (featurep package) (require package nil t)))
```

Comment:
Please refer to https://github.com/emacs-lsp/lsp-mode/pull/1498, this change is undid that.
Is the original commit is not necessary anymore?

Reply:
> It is only slow the first time it is run, because require isn't fast.
> 
> Here is how require works in Emacs:
> There is a set of symbols, called the "features". To check whether a feature is provided (a feature just being a symbol), the `feature` function, implemented in C, checks whether that symbol is in that. A feature can be added to that set using the `provide` function, which is why all modules should end in (provide 'name).
> 
> Require first checks if some feature is provided, and, if not, tries to find the module that it needs to load. This is slow, since it has to check lots of files (the entire `load-path`, 327 directories on my system), but only has to be done once. So (require 'foo) might be somewhat slow the first time, but 'foo will then be provided, so subsequent requires are fast. The issue with lsp-mode was (is?) that some clients which aren't installed are required. This always fails, since those clients aren't in the `load-path`, meaning that those clients never get their features provided, so the `load-path` has to be traversed all the time, which is slow.
> 
> I hope that clears things up.

The slow that #1498 fixed is not from `require` have to look up load path but because of `lsp-register-custom-settings` is called multiple times.
Your change undid that so I would like to ask for more clarification.

Reply:
Okay, `require` seems only load package once, https://www.gnu.org/software/emacs/manual/html_node/elisp/Named-Features.html.
The change in #1498 probably needed to be reworked

Reply:
Actually, since we already guard that, that won't make sense; the features shouldn't be loaded already, meaning optimizing for that case somewhat odd. Should I still revert that?